p2p/src/network/scheduler/p2p_network_scheduler_reducer.rs

```diff
@@ -64,148 +132,480 @@ impl P2pNetworkSchedulerState {
                         limit: P2pNetworkConnectionState::INITIAL_LIMIT,
                     },
                 );
+
+                let dispatcher = state_context.into_dispatcher();
+                dispatcher
+                    .push(P2pNetworkSchedulerEffectfulAction::OutgoingConnect { addr: *addr });
+                Ok(())
             }
-            P2pNetworkSchedulerAction::OutgoingDidConnect { .. } => {
+            P2pNetworkSchedulerAction::OutgoingDidConnect { addr, result } => {
                 // TODO: change to connected
+
+                let (dispatcher, state) = state_context.into_dispatcher_and_state();
+                let p2p_state: &P2pState = state.substate()?;
+
+                match result {
+                    Ok(()) => {
+                        dispatcher.push(P2pNetworkSchedulerEffectfulAction::OutgoingDidConnect {
+                            addr: *addr,
+                        });
+                    }
+                    Err(error) => {
+                        let Some((peer_id, peer_state)) = p2p_state.peer_with_connection(*addr)
+                        else {
+                            bug_condition!(
+                                "outgoing connection to {addr} failed, but there is no peer for it"
+                            );
+                            return Ok(());
+                        };
+                        if matches!(
+                            peer_state.status,
+                            P2pPeerStatus::Connecting(P2pConnectionState::Outgoing(_))
+                        ) {
+                            dispatcher.push(P2pConnectionOutgoingAction::FinalizeError {
+                                peer_id,
+                                error: error.to_string(),
+                            });
+                        }
+                    }
+                }
+                Ok(())
             }
-            P2pNetworkSchedulerAction::IncomingDataIsReady { .. } => {}
             P2pNetworkSchedulerAction::IncomingDataDidReceive { result, addr } => {
-                let Some(state) = self.connections.get_mut(addr) else {
-                    return;
+                if let Some(state) = scheduler_state.connections.get_mut(addr) {
```

Comment:
Maybe add some comment explaining this part since in the original reducer code there was an early return but now that the "effects" code was appended (which is always executed) it is a bit more difficult to follow

Reply:
could also be simplified by moving the logic to a method in `scheduler_state`

Reply:
Since both actions that are dispatcher in "effects" part require connection state, and enabling conditions require connection to exists I will change it so that if connection state is not found to not run anything, and just add comment